### DISCUSSION TITLE
Angular Help (Strange issue when loading massive data from backend in tables)

### DISCUSSION THREAD

**Discussion by ThomasR154** [community/community#138463] [Programming Help] [answered]
### Body

Hello.
I am using Angular, I assume latest version but I am not senior developer.
I wanted to implement data loading in table, like loading limited (50 rows) data in the beginning and load more and more when user reaches the end of scrolling.
I implemented backend apis, spinners etc.
There are no errors but some users say the system crashes strangely.
Here are snippet of my code.

```
import { Component, OnInit, HostListener } from '@angular/core';
import { DataService } from './data.service';

@Component({
  selector: 'app-data-list',
  template: `
    <div *ngFor="let item of items">
      {{ item.id }} - {{ item.name }}
    </div>
  `
})
export class DataListComponent implements OnInit {
  items: any[] = [];
  currentPage: number = 1;
  isLoading: boolean = false;

  constructor(private dataService: DataService) {}

  ngOnInit() {
    this.loadMoreData();
  }

  @HostListener('window:scroll', ['$event'])
  onScroll(event: any) {
    if (
      window.innerHeight + window.scrollY >= document.body.offsetHeight &&
      !this.isLoading
    ) {
      this.loadMoreData();
    }
  }

  loadMoreData() {
    this.isLoading = true;
    this.dataService.getDataPage(this.currentPage).subscribe(
      (data) => {
        this.items = [...this.items, ...data];
        this.currentPage++;
        this.isLoading = false;
      },
      (error) => {
        console.error('Error loading data:', error);
        this.isLoading = false;
      }
    );
  }
}
```

As I repeated, there are no errors, especially on backend and it's working pretty good on my side.
What are the possible issues?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by leoyoung02** [comment, marked as answer]
Hey, how you are doing well.
I assume that's the performance issue.
Here are some solutions that I can think of.




1. Use Virtual Scrolling: Instead of rendering the entire list of items, use a virtual scrolling library like @angular/cdk/scrolling to only render the items that are currently visible on the screen. This can significantly improve the performance of the application, especially when dealing with large datasets.
2. Implement Efficient Change Detection: Optimize the change detection strategy of the component to minimize the number of unnecessary re-renders. You can use the ChangeDetectionStrategy.OnPush strategy and ensure that your component's inputs are immutable to reduce the number of change detection cycles.
3. Debounce Scroll Events: Add a debounce or throttle mechanism to the scroll event handler to reduce the frequency of data loading requests, especially when the user is scrolling rapidly.

**Comment by lqduy** [comment]
# 1. **Virtual Scrolling**

Using virtual scrolling can help reduce the load on the DOM by only rendering the visible items. Angular provides this feature through `@angular/cdk/scrolling`.

Here’s how to implement it:

```bash
npm install @angular/cdk
```

Update your component to use `cdk-virtual-scroll-viewport`:

```html
<cdk-virtual-scroll-viewport itemSize="50" class="viewport">
  <div *cdkVirtualFor="let item of items">
    {{ item.id }} - {{ item.name }}
  </div>
</cdk-virtual-scroll-viewport>
```

Make sure you import the required module in your module file:

```typescript
import { ScrollingModule } from '@angular/cdk/scrolling';

@NgModule({
  imports: [ScrollingModule],
})
export class AppModule {}
```

**Comment by lqduy** [comment]
# 2. **Efficient Change Detection**
Using `ChangeDetectionStrategy.OnPush` can help Angular only check for changes when the inputs have changed. This is especially useful for performance optimization.

In your component, add:

```typescript
import { ChangeDetectionStrategy, Component } from '@angular/core';

@Component({
  selector: 'app-data-list',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <div *ngFor="let item of items">
      {{ item.id }} - {{ item.name }}
    </div>
  `,
})
export class DataListComponent {}
```

For this to work efficiently, ensure that the data coming from the service is immutable, i.e., a new object/array is returned rather than mutating existing data.

**Comment by lqduy** [comment]
# 3. **Debounce Scroll Events**
You can throttle the scroll events to prevent overwhelming the backend with frequent requests as the user scrolls.

To debounce the scroll event, you can use `lodash` or RxJS:

```bash
npm install lodash
```

Then import and use `debounce` in your scroll handler:

```typescript
import { debounce } from 'lodash';

@HostListener('window:scroll', ['$event'])
onScroll(event: any) {
  debounce(() => {
    if (
      window.innerHeight + window.scrollY >= document.body.offsetHeight &&
      !this.isLoading
    ) {
      this.loadMoreData();
    }
  }, 200)();
}
```

Alternatively, you can use RxJS:

```typescript
import { fromEvent } from 'rxjs';
import { debounceTime } from 'rxjs/operators';

ngOnInit() {
  fromEvent(window, 'scroll')
    .pipe(debounceTime(200))
    .subscribe(() => {
      if (
        window.innerHeight + window.scrollY >= document.body.offsetHeight &&
        !this.isLoading
      ) {
        this.loadMoreData();
      }
    });
}
```

**Comment by ThomasR154** [comment]
Oh. That's pretty clear.
I just made some investigation and noticed that I had to use virtual scrolling.
Thanks, @leoyoung02.
Thanks, everybody.
